File: python/p0045.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        # 'jumps' will store the minimum number of jumps required to reach the last index
        jumps = 0

        # 'current_start' and 'current_end' represent the range of indices we can jump from
        current_start = 0
        current_end = 0

        # We loop until 'current_end' reaches the last index of the array
        while current_end < len(nums) - 1:
            # 'farthest_jump' will track the farthest position we can reach within the current jump
            farthest_jump = 0

            # Iterate over all indices from 'current_start' to 'current_end' to find the farthest jump
            for i in range(current_start, current_end + 1):
                # Update the farthest position we can reach from the current position 'i'
                farthest_jump = max(farthest_jump, nums[i] + i)

            # Increase the number of jumps as we are moving to a new range
            jumps += 1

            # Move 'current_start' to the next range (just beyond the current end)
            current_start = current_end + 1

            # Update 'current_end' to the farthest position we can jump to
            current_end = farthest_jump

        # Return the total number of jumps required
        return jumps

File: swift/p0045.swift
class Solution {
    func jump(_ nums: [Int]) -> Int {
        // 'result' will store the minimum number of jumps required to reach the last index
        var jumps = 0

        // 'currentStart' and 'currentEnd' represent the range of indices we can jump from
        var currentStart = 0
        var currentEnd = 0

        // We loop until 'currentEnd' reaches the last index of the array
        while (currentEnd < nums.count - 1) {
            // 'farthestJump' will track the farthest position we can reach within the current jump
            var farthestJump = 0

            // Iterate over all indices from 'currentStart' to 'currentEnd' to find the farthest jump
            for i in currentStart...currentEnd {
                // Update the farthest position we can reach from the current position 'i'
                farthestJump = max(farthestJump, nums[i] + i)
            }

            // Increase the number of jumps as we are moving to a new range
            jumps += 1

            // Move 'currentStart' to the next range (just beyond the current end)
            currentStart = currentEnd + 1

            // Update 'currentEnd' to the farthest position we can jump to
            currentEnd = farthestJump
        }

        // Return the total number of jumps required
        return jumps
    }
}
